docs/snapshotters/erofs.md

```diff
@@ -0,0 +1,143 @@
+# EROFS Snapshotter
+
+The [EROFS](https://erofs.docs.kernel.org) snapshotter is an experimental
+feature, which is able to leverage EROFS-formatted blobs for each committed
+snapshot and prepares an EROFS + OverlayFS mount for each active snapshot.
+
+In order to leverage EROFS-formatted blobs, the EROFS differ is needed to be
+used together to apply image layers.  Otherwise, the EROFS snapshotter will
+just behave as the existing OverlayFS snapshotter: the default applier will
+unpack the image layer into the active EROFS snapshot, and commit it.
+
+Although it sounds somewhat similar to an enhanced OverlayFS snapshotter but
+I believe there are clear differences if looking into `s.mount()` and it highly
+tightens to the EROFS internals.  Currently, it's not quite clear to form an
+enhanced OverlayFS snapshotter directly, and (I think) it's not urgent since
+in the very beginning, it'd be better to be left as an independent snapshotter
+so that existing overlayfs users won't be impacted by the new behaviors and
+users could have a chance to try and develop the related ecosystems (such as
+ComposeFS, confidential containers, gVisor, Kata, gVisor, and more) together.
+
+## Use Cases
+
+The EROFS snapshotter can benefit to several use cases:
+
+For runC containers, instead of unpacking individual files into a directory
+on the backing filesystem, it applies OCI layers into EROFS blobs, therefore:
+
+ - Improved image unpacking performance (~14% for WordPress image with the
+   latest erofs-utils 1.8.2) due to reduced metadata overhead;
+
+ - Full data protection for each snapshot using the S_IMMUTABLE file attribute
+   or fsverity. Currently, fsverity can only protect blob data in the content
+   store;
+
+ - Parallel unpacking can be supported in a more reliable way (fsync) compared
+   to the overlayfs snapshotter (syncfs);
+
+ - Native EROFS layers can be pulled from registries without conversion.
+
+For VM containers, the EROFS snapshotter can efficiently pass through and share
```

Comment:
Given the VM based use-case, it feels to me this has some overlaps with the `blockfile` snapshotter too.

Is there currently a way to get this tried out with Kata based setup?

Reply:
in principle, the blockfile approach is ineffective if underlay fs doesn't support reflink.
Also I'd like to use vhost-user protocol (+ a merged flatten device) to pass through the container image for VM-based containers, rather than generating some loopback device on the host, which is ineffective too.